### PR TITLE
Cloud Native Chapter 6

### PR DESCRIPTION
**Pull request by ardalis**
## Summary

Chapter 6 md files and images

### CODE REVIEW

File: docs/architecture/cloud-native/resilient-communications.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+---
+title: Resilient Communication
+description: Architecting Cloud Native .NET Apps for Azure | Resilient Communication
+ms.date: 06/30/2019
+---
+# Resilient Communications
+
+Throughout this book, we've evangelized the merits of moving beyond traditional monolithic application design and embracing a microservice-based architecture where a set of distributed, self-contained services run independently and communicate with each other using standard communication protocols such as HTTP and HTTPS. While such an architecture buys you many important benefits, it also presents many challenges. Consider for example the following concerns:
+
+- *Out-of-process network communication* Each service communicates over a network protocol which introduces network congestion, latency and transient faults
```
Comment (EugeneKramer) [inline docs/architecture/cloud-native/resilient-communications.md]:
```suggestion
- *Out-of-process network communication* - Each service communicates over a network protocol which introduces network congestion, latency and transient faults.
```

File: docs/architecture/cloud-native/resilient-communications.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+description: Architecting Cloud Native .NET Apps for Azure | Resilient Communication
+ms.date: 06/30/2019
+---
+# Resilient Communications
+
+Throughout this book, we've evangelized the merits of moving beyond traditional monolithic application design and embracing a microservice-based architecture where a set of distributed, self-contained services run independently and communicate with each other using standard communication protocols such as HTTP and HTTPS. While such an architecture buys you many important benefits, it also presents many challenges. Consider for example the following concerns:
+
+- *Out-of-process network communication* Each service communicates over a network protocol which introduces network congestion, latency and transient faults 
+- *Service discovery* - With each service running across a cluster of machines with its own IP address and port, how do services discover and communicate with each other?
+- *Resiliency and failover* How do we manage short-lived failures and keep the system stable?
```
Comment (EugeneKramer) [inline docs/architecture/cloud-native/resilient-communications.md]:
```suggestion
- *Resiliency and failover* - How do we manage short-lived failures and keep the system stable?
```

File: docs/architecture/cloud-native/resilient-communications.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+- *Load balancing* How does inbound traffic get distributed across multiple instances of a service?
```
Comment (EugeneKramer) [inline docs/architecture/cloud-native/resilient-communications.md]:
```suggestion
- *Load balancing* - How does inbound traffic get distributed across multiple instances of a service?
```

File: docs/architecture/cloud-native/resilient-communications.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+- *Security* How are security concerns such as transport-level encryption and certificate management enforced?
```
Comment (EugeneKramer) [inline docs/architecture/cloud-native/resilient-communications.md]:
```suggestion
- *Security* - How are security concerns such as transport-level encryption and certificate management enforced?
```

File: docs/architecture/cloud-native/resilient-communications.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+- *Distributed Monitoring* How do you correlate and capture traceability and monitoring for a single request across multiple consuming services?
```
Comment (EugeneKramer) [inline docs/architecture/cloud-native/resilient-communications.md]:
```suggestion
- *Distributed Monitoring* - How do you correlate and capture traceability and monitoring for a single request across multiple consuming services?
```

**Review by robvet — APPROVED** [review]

File: docs/architecture/cloud-native/application-resiliency-patterns.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+The first line of defense is software-enabled application resiliency. 
+
+While you could invest time writing your own resiliency framework, such frameworks already exist. For example, Polly is a comprehensive .NET resilience and transient-fault-handling library that allows developers to express
+resiliency policies in a fluent and thread-safe manner. Polly targets applications build with both the full .NET and .NET Core Framework. Figure 6-2 the resiliency policies (i.e., functionality) exposed by the Polly Library. These policies can be applied individually or combined together.
+
+![Polly Framework](media/polly-resiliency-framework.png)
+
+**Figure 6-2**. Polly Resiliency Framework Features
+
+Note in the figure above how the various resiliency policies apply to operation requests, whether coming from an external client or another backend service. The goal is to correctly compensate the request for a that be might intermittently unavailable for a short time duration (i.e., in seconds). These interruptions typically manifest themselves with specific HTTP status codes as shown in Figure
```
Comment (mairaw) [inline docs/architecture/cloud-native/application-resiliency-patterns.md]: for a that -> not sure what word you wanted to use here

File: docs/architecture/cloud-native/application-resiliency-patterns.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+6-3.
+
+![HTTP Status Code to Retry](media/http-status-codes.png)
+
+**Figure 6-3**. HTTP Status Code to Retry
+
+Question: Would you retry an HTTP Status Code of 403 - Forbidden? No. In that case, the system is functioning properly and telling you are simply not authorized to perform the operation that you are requesting. Care must be taken in selecting appropriate operations to retry.
+
+As recommended in Chapter 1, Microsoft developers targeting cloud native applications should be building on the .NET Core framework. At the time of this writing, version 2.2 is the current release. Version 2.1 introduced the HTTPClientFactory library as a for sending HTTP requests and receiving HTTP responses from a resource identified as a URL. Superseding the original HTTPClient class, the factory class supports many enhanced features, one of which is tight integration with the Polly resiliency framework. It enables you to easily define resiliency policies in the application Startup class to handle partial failures and connectivity issues. Figure X below shows a code block that configures resiliency for a given service.
```
Comment (mairaw) [inline docs/architecture/cloud-native/application-resiliency-patterns.md]: as a for sending -> as a way for sending?

File: docs/architecture/cloud-native/infrastructure-resiliency-azure.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+![Scale Up vs. Scale Out](media/scale-up-scale-out.png)
+
+**Figure 6-8.** Scale Up vs. Scale Out
+
+- *Scale proportionally.* When scaling a service, think in terms of *resource sets*. If you were to dramatically scale out a specific service, what impact would that have on backend data stores, caches and dependent services? Some resources such as Cosmos DB can scale out proportionally, while many other    cannot. You want to ensure that you do not scale out a resource that will to a point where it will choke other associated resources.
```
Comment (mairaw) [inline docs/architecture/cloud-native/infrastructure-resiliency-azure.md]: will to a -> will what?

File: docs/architecture/cloud-native/infrastructure-resiliency-azure.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+- *Avoid affinity.* A best practice is to ensure a node doesn't require local affinity, often referred to as a *sticky session*. A request should be able to route to any instance. If you do need to persist state, it should be saved to a distributed cache, such as Azure Redis cache.
+
+- *Take advantage of platform autoscaling features.* Use built-in autoscaling features whenever possible, rather than custom or third-party mechanisms. Use scheduled scaling rules, where possible, to ensure that resources are available without a startup delay, but add reactive autoscaling to the rules, where appropriate, to cope with unexpected changes in demand. For more information, see Autoscaling guidance.
+
+- *Scale-up aggressively.* A final practice would be to scale up aggressively so that you can quickly meet immediate spikes in traffic without losing business. And, then scale down (remove unneeded resources) conservatory in order to keep the system stable. A simple way to implement this might be to set the cool down period (the time to wait between scaling operations) for adding resources to 5 minutes and the period for removing instances up to 30 minutes.
```
Comment (mairaw) [inline docs/architecture/cloud-native/infrastructure-resiliency-azure.md]: conservatory -> conservatively?

File: docs/architecture/cloud-native/infrastructure-resiliency-azure.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+**Figure 6-8.** Scale Up vs. Scale Out
+
+- *Scale proportionally.* When scaling a service, think in terms of *resource sets*. If you were to dramatically scale out a specific service, what impact would that have on backend data stores, caches and dependent services? Some resources such as Cosmos DB can scale out proportionally, while many other    cannot. You want to ensure that you do not scale out a resource that will to a point where it will choke other associated resources.
+
+- *Avoid affinity.* A best practice is to ensure a node doesn't require local affinity, often referred to as a *sticky session*. A request should be able to route to any instance. If you do need to persist state, it should be saved to a distributed cache, such as Azure Redis cache.
+
+- *Take advantage of platform autoscaling features.* Use built-in autoscaling features whenever possible, rather than custom or third-party mechanisms. Use scheduled scaling rules, where possible, to ensure that resources are available without a startup delay, but add reactive autoscaling to the rules, where appropriate, to cope with unexpected changes in demand. For more information, see Autoscaling guidance.
```
Comment (mairaw) [inline docs/architecture/cloud-native/infrastructure-resiliency-azure.md]: missing link

File: docs/architecture/cloud-native/infrastructure-resiliency-azure.md [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+### Built-in Retry in services
+
+We encouraged the best practice of implementing programmatic retry operations in an earlier section. Keep in mind that many Azure services and their corresponding client SDKs also include retry mechanisms. Here, we summarize retry features in the many of the Azure services we have discussed in this book.
+
+- *Azure Cosmos DB.* The DocumentClient class from the client API automatically retires failed attempts. The number of retries and maximum wait time are configurable. Exception thrown by the client API are either requests that exceed the retry policy or non-transient errors.
+
+- *Azure Redis Cache.* The Redis StackExchange client uses a connection manager class that includes retries on failed attempts. The number of retries, specific retry policy and wait time are all configurable.
+
+- *Azure Service Bus.* The Service Bus client exposes a RetryPolicy property that can be configured with a back-off interval, retry count and TerminationTimeBuffer which specifies the maximum time an operation can take. The default policy is 9 maximum retry attempts with a 30 second backoff period between attempts.
```
Comment (mairaw) [inline docs/architecture/cloud-native/infrastructure-resiliency-azure.md]: RetryPolicy property or class (https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.retrypolicy?view=azure-dotnet)?

File: docs/architecture/cloud-native/resiliency.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+Therefore, your application and infrastructure must be resilient. In the following sections we'll explore both application and cloud infrastructure resiliency.
+
+Resiliency is the ability to for your system to react to problems and still remain functional. It isn't about avoiding failures but accepting the fact that failures are inevitable and responding to them in a way that avoids downtime or data loss. The goal of resiliency is to return the application to a fully functioning state after a failure.
+
+As discussed in Chapter 1, cloud native systems embrace distributed architecture: Independent microservices interacting with cloud-based [backing services](https://12factor.net/backing-services) communicating over a network layer. 
+
+It's challenging to design and deploy a distributed cloud-native application. And, it's equally challenging to keep that application running in an environment where failure is certain. Consider what can go wrong:
+
+- Unexpected [Network Latency](https://www.techopedia.com/definition/8553/network-latency) network latency
```
Comment (mairaw) [inline docs/architecture/cloud-native/resiliency.md]: this is repeated. not sure which one you wanna keep.

**Comment by mairaw** [comment]
Review is done @ardalis. But there are some pending comments that we should resolve before merging.

**Review by mairaw — CHANGES_REQUESTED** [review]
Please look at my questions before we can merge

**Comment by ardalis** [comment]
Should be good to merge, @mairaw

**Comment by ardalis** [comment]
This one says there's 1 change requested but I don't see what it is.

**Comment by mairaw** [comment]
We were having some build issues earlier today. Trying again.

File: docs/architecture/cloud-native/infrastructure-resiliency-azure.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+- *Azure Cosmos DB.* The <xref:Microsoft.Azure.Documents.Client.DocumentClient> class from the client API automatically retires failed attempts. The number of retries and maximum wait time are configurable. Exceptions thrown by the client API are either requests that exceed the retry policy or non-transient errors.
+
+- *Azure Redis Cache.* The Redis StackExchange client uses a connection manager class that includes retries on failed attempts. The number of retries, specific retry policy and wait time are all configurable.
+
+- *Azure Service Bus.* The Service Bus client exposes a [RetryPolicy property](https://docs.microsoft.com/dotnet/api/microsoft.servicebus.retrypolicy?view=azure-dotnet) that can be configured with a back-off interval, retry count, and <xref:Microsoft.ServiceBus.RetryExponential.TerminationTimeBuffer>, which specifies the maximum time an operation can take. The default policy is nine maximum retry attempts with a 30-second backoff period between attempts.
```
Comment (mairaw) [inline docs/architecture/cloud-native/infrastructure-resiliency-azure.md]: If that's really the API you were talking about, this is a class

```suggestion
- *Azure Service Bus.* The Service Bus client exposes a [RetryPolicy class](xref:Microsoft.ServiceBus.RetryPolicy) that can be configured with a back-off interval, retry count, and <xref:Microsoft.ServiceBus.RetryExponential.TerminationTimeBuffer>, which specifies the maximum time an operation can take. The default policy is nine maximum retry attempts with a 30-second backoff period between attempts.
```

**Comment by mairaw** [comment]
Edited the last one that had major issues before. One comment left.

**Comment by ardalis** [comment]
Hopefully good to go now.

**Review by mairaw — APPROVED** [review]

**Pull request merged**